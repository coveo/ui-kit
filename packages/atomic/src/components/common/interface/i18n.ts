import {getAssetPath} from '@stencil/core';
import DOMPurify from 'dompurify';
import Backend, {HttpBackendOptions} from 'i18next-http-backend';
import availableLocales from '../../../generated/availableLocales.json';
import {AnyEngineType} from './bindings';
import {BaseAtomicInterface} from './interface-common';

export const i18nTranslationNamespace = 'translation';

function isI18nLocaleAvailable(locale: string) {
  return availableLocales.includes(locale.toLowerCase());
}

export function i18nBackendOptions(
  atomicInterface: BaseAtomicInterface<AnyEngineType>
): HttpBackendOptions {
  return {
    loadPath: `${getAssetPath(
      atomicInterface.languageAssetsPath
    )}/{{lng}}.json?lng={{lng}}&ns={{ns}}`,
    request: async (_options, url, _payload, callback) => {
      try {
        const [fetchUrl, searchParams] = url.split('?');
        const urlParams = new URLSearchParams(searchParams);
        const lng = urlParams.get('lng')!;
        const ns = urlParams.get('ns')!;

        if (!isI18nLocaleAvailable(lng)) {
          throw new Error(`Unsupported locale "${lng}"`);
        }

        if (ns !== i18nTranslationNamespace) {
          throw new Error(`Unsupported namespace "${ns}"`);
        }

        const response = await fetch(fetchUrl);
        if (response.status !== 200 && response.status !== 304) {
          throw new Error(
            `Unsuccessful request returned status "${response.status}"`
          );
        }

        callback(null, {
          status: response.status,
          data: await response.json(),
        });
      } catch (error) {
        callback(error, {status: 404, data: ''});
      }
    },
  };
}

export function init18n(atomicInterface: BaseAtomicInterface<AnyEngineType>) {
  return atomicInterface.i18n.use(Backend).init({
    debug: atomicInterface.logLevel === 'debug',
    lng: atomicInterface.language,
    nsSeparator: '___',
    fallbackLng: 'en',
    backend: i18nBackendOptions(atomicInterface),
    interpolation: {
      escape: (str) => DOMPurify.sanitize(str),
    },
    compatibilityJSON: 'v3',
  });
}

declare module 'i18next' {
  interface CustomTypeOptions {
    resources: Record<string, Record<string, string>>;
  }
}
