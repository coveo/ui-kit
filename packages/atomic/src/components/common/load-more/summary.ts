import {i18n} from 'i18next';
import {html} from 'lit';
import {unsafeHTML} from 'lit/directives/unsafe-html.js';
import {ItemKind} from './common-types';

type LoadMoreSummaryLabel = `showing-${ItemKind}-of-load-more`;

interface LoadMoreSummaryProps {
  i18n: i18n;
  from: number;
  to: number;
  label: LoadMoreSummaryLabel;
}

export const loadMoreSummary = ({
  i18n,
  from,
  to,
  label,
}: LoadMoreSummaryProps) => {
  const wrapHighlight = (content: string) => {
    return `<span class="font-bold text-on-background" part="highlight">${content}</span>`;
  };

  const locale = i18n.language;
  const content = i18n.t(label, {
    interpolation: {escapeValue: false},
    last: wrapHighlight(from.toLocaleString(locale)),
    total: wrapHighlight(to.toLocaleString(locale)),
  });

  return html`
    <div class="text-neutral-dark my-2 text-lg" part="showing-results">
      ${unsafeHTML(content)}
    </div>
  `;
};
