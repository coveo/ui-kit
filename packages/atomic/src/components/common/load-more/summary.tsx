import {FunctionalComponent, h} from '@stencil/core';
import {i18n} from 'i18next';

interface LoadMoreSummaryProps {
  i18n: i18n;
  from: number;
  to: number;
}

export const LoadMoreSummary: FunctionalComponent<LoadMoreSummaryProps> = ({
  i18n,
  from,
  to,
}) => {
  const wrapHighlight = (content: string) => {
    return `<span class="font-bold text-on-background" part="highlight">${content}</span>`;
  };

  const locale = i18n.language;
  const content = i18n.t('showing-results-of-load-more', {
    interpolation: {escapeValue: false},
    last: wrapHighlight(from.toLocaleString(locale)),
    total: wrapHighlight(to.toLocaleString(locale)),
  });

  return (
    <div
      class="my-2 text-lg text-neutral-dark"
      part="showing-results"
      innerHTML={content}
    ></div>
  );
};
