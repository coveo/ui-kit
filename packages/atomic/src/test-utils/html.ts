export function getInnerHTMLWithoutComments(element: Element): string {
  return element.innerHTML.replaceAll(/<!--.*?-->/gs, '');
}
