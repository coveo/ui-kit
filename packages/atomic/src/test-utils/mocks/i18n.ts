import enTranslations from '@/dist/atomic/lang/en.json';
import i18next from 'i18next';

export async function getI18nTestInstance() {
  const i18n = i18next.createInstance();
  await i18n.init({
    lng: 'en',
    resources: {
      en: {
        translation: enTranslations,
      },
    },
  });
  return i18n;
}
