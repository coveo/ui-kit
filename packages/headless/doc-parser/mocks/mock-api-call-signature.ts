import {
  ApiCallSignature,
  IApiCallSignatureOptions,
  ReleaseTag,
} from '@microsoft/api-extractor-model';

export function buildMockApiCallSignature(
  config: Partial<IApiCallSignatureOptions> = {}
) {
  return new ApiCallSignature({
    docComment: undefined,
    excerptTokens: [],
    releaseTag: ReleaseTag.None,
    overloadIndex: 0,
    parameters: [],
    returnTypeTokenRange: {startIndex: 0, endIndex: 0},
    typeParameters: [],
    ...config,
  });
}
