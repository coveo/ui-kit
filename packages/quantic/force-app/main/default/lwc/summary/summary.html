<template>
  <template if:false={hasResults}>
    <p class="slds-box slds-theme_alert-texture slds-m-top_medium">No results<span class="bold">{forQuery}</span></p>
  </template>
  <template if:true={hasResults}>
    <p class="slds-text-heading_small slds-m-top_medium">
      Results<span class="bold">{range}</span>
      <span class="bold">{total}</span>
      <span class="bold">{forQuery}</span>{duration}
    </p>
  </template>
</template>
