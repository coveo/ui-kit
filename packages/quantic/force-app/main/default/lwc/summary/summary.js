import {LightningElement, track, api} from 'lwc';
import {registerComponentForInit, initializeWithHeadless} from 'c/headlessLoader';

export default class Summary extends LightningElement {
  @track state = {};

  /** @type {import("coveo").QuerySummary} */
  querySummary;
  /** @type {import("coveo").Unsubscribe} */
  unsubscribe;

  constructor() {
    super();
    registerComponentForInit(this);
  }

  connectedCallback() {
    initializeWithHeadless(this, this.initialize.bind(this));
  }

  /**
   * @param {import("coveo").Engine} engine
   */
  @api
  initialize(engine) {
    this.querySummary = CoveoHeadless.buildQuerySummary(engine);
    this.unsubscribe = this.querySummary.subscribe(() => this.updateState());
  }

  disconnectedCallback() {
    if (this.unsubscribe) {
      this.unsubscribe();
    }
  }
  updateState() {
    this.state = this.querySummary.state;
  }

  get hasResults() {
    return this.state.hasResults;
  }

  get hasQuery() {
    return this.state.hasQuery;
  }

  get forQuery() {
    return this.state.hasQuery ? ` for ${this.state.query}` : '';
  }

  get range() {
    return ` ${this.state.firstResult} - ${this.state.lastResult}`;
  }

  get total() {
    return ` of ${this.state.total.toString()}`;
  }

  get duration() {
    if (this.state.hasDuration) {
      return ` in ${this.state.durationInSeconds} seconds`;
    }

    return '';
  }
}
