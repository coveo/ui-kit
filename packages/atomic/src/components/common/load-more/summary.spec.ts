import {getInnerHTMLWithoutComments} from '@/test-utils/html';
import {getI18nTestInstance} from '@/test-utils/mocks/i18n';
import type {i18n as I18n} from 'i18next';
import {html, render} from 'lit';
import {loadMoreSummary} from './summary';

describe('loadMoreSummary', () => {
  let i18n: I18n;
  let container: HTMLElement;

  beforeAll(async () => {
    i18n = await getI18nTestInstance();
  });

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  const renderLoadMoreSummary = (
    props: Parameters<typeof loadMoreSummary>[0]
  ) => {
    render(html`${loadMoreSummary(props)}`, container);
  };

  it('should render the correct HTML structure', () => {
    renderLoadMoreSummary({
      i18n,
      from: 1,
      to: 10,
      label: 'showing-results-of-load-more',
    });

    expect(getInnerHTMLWithoutComments(container)).toMatchSnapshot();
  });

  it('should localize the content correctly', () => {
    renderLoadMoreSummary({
      i18n,
      from: 1,
      to: 10,
      label: 'showing-results-of-load-more',
    });
    expect(container).toHaveTextContent('Showing 1 of 10 results');
  });

  it('should bolden the numbers', () => {
    renderLoadMoreSummary({
      i18n,
      from: 1,
      to: 10,
      label: 'showing-results-of-load-more',
    });
    const highlights = container.querySelectorAll('[part="highlight"]');

    expect(highlights).toHaveLength(2);
    expect(highlights[0]).toHaveTextContent('1');
    expect(highlights[1]).toHaveTextContent('10');
    expect(highlights[0].className).toBe(highlights[1].className);
    expect(highlights[0].classList).toContain('font-bold');
  });
});
