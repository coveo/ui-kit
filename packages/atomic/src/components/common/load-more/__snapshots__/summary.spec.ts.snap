// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadMoreSummary > should render the correct HTML structure 1`] = `
"
    <div class="text-neutral-dark my-2 text-lg" part="showing-results">
      Showing <span class="font-bold text-on-background" part="highlight">1</span> of <span class="font-bold text-on-background" part="highlight">10</span> results
    </div>
  "
`;
